doc/Understanding_Layout_Schema.rst

```diff
@@ -0,0 +1,567 @@
+
+Lustre has two levels of mapping. The first level of mapping is stored on MDS as a file EA. The layout is structured as {{ost1,fid1}, {ost2,fid2}, {ost3,fid3}, {ost4,fid4}, … {ost-n,fid-n}}, along with stripe size. So the logical offset in a file can be calculated into one of the object represented by {ost-index, fid}. With this info, lustre client communicates to that ost, accessing the object with that specified fid. Another mapping from object logical offset to disk block is done by a local disk file system, ldiskfs (the origin of ext4).
+
+pNFS and file layouts
+=====================
+
+In a traditional NAS deployment, there is a NAS server and multiple NFS or CIFS clients. With the advent of SAN, even the client machines started having direct access to the storage. To boost the performance of overall system, pNFS  like systems were developed. The pNFS  clients started requesting file layouts so that they could perform direct I/O (this does not mean un buffered I/O) to the files. This created different kinds of requirements for the file-layouts. The layouts that are to be transported need to be  
+
+- Compact (not occupying a lot of space/memory) 
+
+- Possibly they need to be in network-byte order (pre-encoded) for easy transport 
+
+- A part (region of a file) of the layout could be delivered if necessary 
+
+***************
+Design Goals
+***************
+
+Looking at the history of the file-layouts, it’s very clear that the file layouts evolve over a period of time. A host of copy services (snapshot, back-up, archiving, data de-duplication, replication etc) can also influence the file layout. In Motr the layout will be transported between two servers. The file layout will need a compact description so that it will consume minimal resources (memory, network bandwidth). Motr should support multiple layouts. It should allow accommodation of new (future) layouts. It should allow migration of file data from one layout to another. The design goals for the file layouts (and schema) are listed below:
+
+- Support multiple layouts within the file-system 
+
+- Provide support to add new layouts 
+
+- Provide support for composite/mixed layouts (combining different layout types together) 
+
+- Provide migration of file data from one layout to another 
+
+- Provide compact description of the layouts 
+
+- Provide pre-encoded layouts (where necessary) for easy transport 
+
+*********************
+Motr Layout Types  
+*********************
+
+HLD for layout schema lists the following layout types: 
+
+- SNS 
+
+- Local RAID 
+
+- RAID levels 
+
+- Future layouts such as de-duplication, encryption, compression 
+
+Although these layout types are well classified from schema perspective, we believe there will be two types of layouts that will be supported for the planned demo: 
+
+- Parity de-clustering layout (c2_pdclust_layout) (Conceptually this is ‘SNS layout with parity de-clustering as a property’ and is referred as pdclust layout.) 
```

Comment:
Please replace the "c2_" prefix with "m0_" prefix.

Reply:
Incorporated.